### DISCUSSION TITLE
how to avoid vitest to complain about unhandled promise-rejection

### DISCUSSION THREAD

**Discussion by rburgst** [vitest-dev/vitest#3689] [Q&A] [answered]
I have a test that tests an error case where a method which uses `setTimeout` will throw an error. 
This causes vitest to return a non-zero exit code.

As this is expected behaviour (i.e. the test specifically tests this behaviour), I dont want to turn off unhandled promise rejections globally.
I already tried using `vi.runAllTimers` etc but I cannot ignore the promise rejection for that individual test.

here is a sample 
```typescript
    it('should not report unhandled promise rejections', async () => {
        const func = vi.fn();
        func.mockRejectedValueOnce(new Error('test error'));

        setTimeout(() => {
            func();
        }, 100);
        await vi.runAllTimersAsync();
    });
```

**Comment by KatoakDR** [comment, marked as answer]
Hello, per the sample test you provided. The problem is that `func` _is a promise_ and must be `await`ed or part of a `.then().catch()` chain for the error to be handled properly.

Since `setTimeout` does not accept promises, you need to use the `.then().catch()` chain approach.

Below I've provided an updated version of your sample test using two spies to assert if the `func` promise resolves or rejects.

```ts
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';

describe('async-timeout-test', () => {
  beforeEach(() => {
    vi.useFakeTimers({ shouldAdvanceTime: true });
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('should not report unhandled promise rejections', async () => {
    const resolveSpy = vi.fn();
    const rejectSpy = vi.fn();

    const func = vi.fn();
    func.mockRejectedValueOnce(new Error('test error'));

    setTimeout(() => {
      func()
        .then(() => {
          resolveSpy();
        })
        .catch((error: unknown) => {
          rejectSpy(error);
        });
    }, 100);

    await vi.runAllTimersAsync();

    expect(resolveSpy).not.toHaveBeenCalled();
    expect(rejectSpy).toHaveBeenCalledWith(new Error('test error'));
  });
});
```

**Reply by rburgst** [reply]
Thanks for the suggestion. This is a bit involved and leads to hard to read tests unfortunately. It would be nicer if there was a helper for vitest that simplifies such a case.
